File: 15. 3Sum.py
class Solution(object):
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        res = []
        ln = len(nums)
        nums.sort()
        i,j,k = 0,0,0
        while i < ln:
            l, h = i+1, ln-1
            while l < h:
                val = nums[i] + nums[l] + nums[h]                
                if val > 0:
                    h -= 1
                    continue
                if val < 0:
                    l += 1
                    continue

                if val == 0:
                    res.append([nums[i], nums[l], nums[h]])

                while l < h and nums[l] == nums[l + 1]:
                    l += 1
                l += 1

            while i < ln - 1 and nums[i] == nums[i+1]:
                i += 1
            i += 1
        
        return res

s = Solution()
print s.threeSum([0,0,0,0])

File: go/src/leetcode/15. 3Sum/15. 3Sum.go
package main

import (
	"fmt"
	"sort"
)

func threeSum(nums []int) [][]int {
	ans := [][]int{}
	n := len(nums)
	sort.Ints(nums)
	for i := 0; i < n; i++ {
		l, r := i+1, n-1
		for l < r {
			val := nums[i] + nums[l] + nums[r]
			if val < 0 {
				l++
			} else if val > 0 {
				r--
			} else {
				ans = append(ans, []int{nums[i], nums[l], nums[r]})
				for l < r && nums[l] == nums[l+1] {
					l++
				}
				l++
			}
		}
		for ;i < n-1 && nums[i] == nums[i+1]; i++ {}
	}

	return ans
}

func main() {
	fmt.Println(threeSum([]int{-1,0,1,2,-1,-4}))
	fmt.Println(threeSum([]int{}))
	fmt.Println(threeSum([]int{0}))
}
